src/relay/backend/vm/lambda_lift.cc

```diff
@@ -0,0 +1,185 @@
+        }
+
+        auto free_vars = FreeVars(func);
+        auto free_type_vars = FreeTypeVars(func, module_);
+        auto body = Downcast<Function>(ExprMutator::VisitExpr_(func_node));
+
+        // When performing this optimization there are two
+        // cases.
+        //
+        // The first case in which we have no free variables
+        // we can just lift the function into the global
+        // environment without needing to allocate a closure.
+        //
+        //
+        // The second case requires that we generate a special
+        // function with makes a distinction between allocating
+        // a closure, and then the code for the closure.
+        //
+        // We represent a closure allocation by lifting the
+        // closure to a global function which takes its
+        // captured arguments and then directly returns
+        // the function representing the closure's code.
+        //
+        // When we generate code later on a call to the "outer"
+        // function marked as a closure is used to emit allocation
+        // code for the closure's environment.
+        //
+        // The "inner" function is should be used to generate the
+        // code for the closure.
+        Function lifted_func;
+        if (free_vars.size() == 0) {
+          lifted_func =
+            FunctionNode::make(
+              body->params,
+              body->body,
+              body->ret_type,
+              free_type_vars);
+        } else {
+          lifted_func =
+            FunctionNode::make(
+                free_vars,
+                body,
+                func->func_type_annotation(),
+                free_type_vars);
+
+          lifted_func = MarkClosure(lifted_func);
+        }
+
+        CHECK(lifted_func.defined());
+
+        auto name = GenerateName(lifted_func);
+        auto global = this->module_->GetGlobalVar(name);
+
+        lifted_.push_back({global, lifted_func});
+
+        if (free_vars.size() == 0) {
+          return std::move(global);
+        } else {
+          // If we need to allocate a closure
+          // we pass the variables in its environment
+          // here.
+          Array<Expr> fvs;
+          for (auto fv : free_vars) {
+            fvs.push_back(fv);
+          }
+          return CallNode::make(global, fvs);
+        }
+    }
+
+    Function Lift(const Function& func) {
+        RELAY_LOG(INFO)
+          << "Lifting: "
+          << AsText(func, false)
+          << std::endl;
+        return FunctionNode::make(
+            func->params,
+            VisitExpr(func->body),
+            func->ret_type,
+            func->type_params,
+            func->attrs);
+    }
+};
+
+/* The goal of this pass is to lift out any nested functions into top-level
+ * functions.
+ *
+ * We will lift the functions out into globals which take the set of the free vars
+ * and then return a function whcih has b
+ */
+Module LambdaLift(const Module& module)  {
```

Comment:
if it is module->module, perhaps it should be a pass instead of vm

Reply:
I have kept all the VM compiler passes by themselves for now. I do not want to expose them in the main pass header until we move to the pass manager.